### PR TITLE
added cl_half support for test_geometric

### PR DESCRIPTION
**Pull request by shajder**
According to issue description (geometrics section):

https://github.com/KhronosGroup/OpenCL-CTS/issues/142

Couple of additional remarks for reviewers:

- Source files `test_geometrics.cpp` and `test_geometrics_double.cpp` were unified in the process of adding `cl_khr_fp16` support. I've tried to make as little corrections to original code as possible. As result we get quite a large source file which could be easily divided into 3-4 logic parts - I left it as is ATM for the sake of comparison purposes.

- Please take a closer look at `GeometricsFPTest::FillWithTrickyNums`, former `FillWithTrickyNumbers`/`FillWithTrickyNumbers_double` since the logic behind wasn't entirely clear for me. Moreover it crashed after decreasing `TEST_SIZE` macro. I would gladly refactor this method to make it a bit more readable.

- BTW. I've decreased `TEST_SIZE` from `1<<20` to `1<<10`, please let me know if you think it's not enough. trickyNumbers were copied into max 320 elements, rest of it equals 0 or random value which seems like sufficient number of tests.

- Please take a closer look at `dot` verification code for halfs - it was source of pain (`TwoArgsFPTest::TwoArgsKernel`). I had to apply additional condition to test expected/reference values against overflow/underflow.

- Please take a look at verification of ULP limits - previous calculations seemed not entirely related to spec so I've applied my corrections (still it makes me think I may be mistaken here).

### CODE REVIEW

File: test_conformance/geometrics/test_geometrics_fp.cpp [hunk 1]
```diff
@@ -0,0 +1,2240 @@
...
+            std::stringstream sstr;
+            std::string printout;
+            if (ulpLimit < 0)
+            {
+                // Limit below zero means we need to test via a computed error
+                // (like cross product does)
+                T maxValue = GetMaxValue<T>(inDataA + i * vecSize,
+                                            inDataB + i * vecSize, vecSize, p);
+
+                double error = 0.0, errorTolerance = 0.0;
```
Comment (EwanC) [inline test_conformance/geometrics/test_geometrics_fp.cpp:1384]: I'm seeing errors for cross and dot tests. The output from the verification error itself is informative, so nice work there, but as an aside the `CL_DEVICE_NOT_FOUND` in the backtrace is slightly confusion.

cross fail
```
ERROR: Data sample 122 does not validate! Expected (inf,inf,-0x1.b28p+15,0x0p+0), got (0x0p+0,0x1.0040140100001p+2,0x0p+0,0x1.0040140100401p+2)
    Input: (0x1.cbp+7 -0x1.9d8p+6 0x1.9dcp+8) and (0x1.824p+8 -0x1.a04p+8 -0x1.2dcp+6)
    Errors: (nan out of 0x1.fb9c0cp+8), (nan out of 0x1.f587ccp+8), (inf out of 0x1.fb9c0cp+8)
     ulp 0.000000
ERROR: CrossFPTest::RunSingleTest: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:1012)
ERROR: GeometricsFPTest::Run: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:933)
ERROR: Test Failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.h:325)
geom_cross FAILED
```

dot failed
```
ERROR: Data sample 122 at size 2 does not validate! Expected (-0x1.03b78p+14), got (-inf), sources (0x1.8dp+8 and -0x1.7acp+7) error of inf against tolerance 499.109
        vector A: { 397, 141.875,}, vector B: { -189.375, 412.75,}   geom_two_args vector size 2 FAILED
ERROR: TwoArgsFPTest::RunSingleTest: test_geometrics failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:1175)
ERROR: GeometricsFPTest::Run: test_relational failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.cpp:933)
ERROR: Test Failed! (CL_DEVICE_NOT_FOUND from ../test_conformance/geometrics/test_geometrics_fp.h:325)
geom_dot FAILED
```

From debugging this and why our internal testing didn't pick it up, I interpreted the spec absolute error calculations as being done in half precision. This meant that `'max * max * (2n - 1) * HLF_EPSILON'`
could overflow in an intermediate calculation. Essentially giving us
more leeway for our dot implementation to overflow.

I think we should have a spec clarification for all of half, float, and double.
dot/crosss absolute errors about what precision that error is calculated in. A larger precision than what's being tested makes sense, to match the wording from ULP - "the reference value used to compute the ULP value of an arithmetic operation is the infinitely precise result"
which is what this PR does already. If so, I think it is our implementation that should be improved.

TLDR - I think this PR checks dot/cross error correctly, but we could use a spec clarification about the precision those absolute errors are calculated in.

**Review by EwanC — COMMENTED** [review]
Need to review this more thoroughly, but overall design is certainly an improvement.

File: test_conformance/geometrics/test_geometrics_fp.cpp [hunk 2]
```diff
@@ -0,0 +1,2240 @@
...
+            break;
+        default: test_error(-1, "LengthFPTest::Run: incorrect fp type"); break;
+    }
+    test_error(error, "LengthFPTest::RunSingleTest: test_geometrics failed");
+    return CL_SUCCESS;
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T> int LengthFPTest::LenghtTest(OneArgTestParams<T> &p)
```
Comment (EwanC) [inline test_conformance/geometrics/test_geometrics_fp.cpp:1723]: `LenghtTest` -> `LengthTest`

File: test_conformance/geometrics/test_geometrics_fp.cpp [hunk 3]
```diff
@@ -0,0 +1,2240 @@
+//
```
Comment (EwanC) [inline test_conformance/geometrics/test_geometrics_fp.cpp:1]: This new file is pretty monolithic, although it does solve the previous code duplication.

Is there any way to break it down into separate files to make it the code easier to navigate? The template instantiation
might be problematic. Thinking splitting out common/helper code, or verification code, or setup code.

Comment (shajder) [reply test_conformance/geometrics/test_geometrics_fp.cpp:1]: Absolutely. I see four logical parts of this test and my proposition is to create separate source files:

1) Cross test
2) Dot / FastDistance / Distance tests
3) Length / FastLength
4) Normalize / FastNormalize

... plus header file for base structures. As visible, we could divide even more if you think that could impact readability of the code.

Comment (EwanC) [reply test_conformance/geometrics/test_geometrics_fp.cpp:1]: That sounds like a good organization to me, I don't think we need to go more granular than your proposal.

File: test_conformance/geometrics/test_geometrics_fp.cpp (hunk 2 above)
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_fp.cpp]: Done

File: test_conformance/geometrics/test_geometrics_base.h [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+#include <map>
+#include <memory>
+#include <CL/cl_half.h>
+
+#include "testBase.h"
+#include <limits>
+#include <cmath>
+
+#define HALF_P_NAN 0x7e00
+#define HALF_N_NAN 0xfe00
```
Comment (EwanC) [inline test_conformance/geometrics/test_geometrics_base.h:30]: A quick search doesn't show these NAN values used anywhere

File: test_conformance/geometrics/test_geometrics_base.h [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+#define HALF_P_INF 0x7c00
```
Comment (EwanC) [inline test_conformance/geometrics/test_geometrics_base.h:31]: These INF values are only used in 1 pace, so might be easier to just inline them

File: test_conformance/geometrics/test_geometrics_base.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            strcpy(GeomTestBase::extension,
+                   "#pragma OPENCL EXTENSION cl_khr_fp64 : enable\n");
+        else if (p->dataType == kHalf)
+            strcpy(GeomTestBase::extension,
+                   "#pragma OPENCL EXTENSION cl_khr_fp16 : enable\n");
+        else
+            GeomTestBase::extension[0] = '\0';
+
+        error = RunSingleTest(p.get());
+        test_error(error, "GeometricsFPTest::Run: test_relational failed");
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_base.cpp]: Done

File: test_conformance/geometrics/test_geometrics_cross.cpp [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+            test_error(-1, "CrossFPTest::RunSingleTest: incorrect fp type");
+            break;
+    }
+    test_error(error, "CrossFPTest::RunSingleTest: test_relational failed");
+    return CL_SUCCESS;
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T> int CrossFPTest::CrossKernel(const GeomTestParams<T> &p)
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_cross.cpp]: Done

File: test_conformance/geometrics/test_geometrics_base.h [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+                                          const GeomTestParams<T> &p)
+{
+    const size_t trickyCount = p.trickyValues.size();
+    size_t copySize = vecSize * vecSize * trickyCount;
+
+    if (copySize * 2 > num_elems)
+    {
+        log_error(
+            "Buffer size not sufficient for test purposes, please verify!");
+        return;
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_base.h]: Done

File: test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+    }
+    test_error(error,
+               "TwoVecToScalarFPTest::RunSingleTest: test_geometrics failed");
+    return CL_SUCCESS;
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T>
+int TwoVecToScalarFPTest::TwoVecToScalar(const TwoVecToScalarTestParams<T> &p)
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp]: Done

File: test_conformance/geometrics/test_geometrics_base.h (hunk 4 above)
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_base.h]: Done

File: test_conformance/geometrics/test_geometrics_base.h (hunk 5 above)
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_base.h]: Done

File: test_conformance/geometrics/test_geometrics_vec_to_vec.cpp [hunk 10]
```diff
@@ -0,0 +1,630 @@
...
+        }
+    }
+    return CL_SUCCESS;
+}
+
+//--------------------------------------------------------------------------
+
+cl_int FastNormalizeFPTest::SetUp(int elements)
+{
+    // only float supports fast_distance
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_vec_to_vec.cpp]: done

File: test_conformance/geometrics/test_geometrics_vec_to_scalar.cpp [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+        }
+    }
+    return CL_SUCCESS;
+}
+
+//--------------------------------------------------------------------------
+
+cl_int FastLengthFPTest::SetUp(int elements)
+{
+    // only float supports fast_distance
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_vec_to_scalar.cpp]: done

File: test_conformance/geometrics/test_geometrics_vec_to_vec.cpp [hunk 12]
```diff
@@ -0,0 +1,630 @@
...
+template <typename T>
+int NormalizeFPTest::NormalizeTest(VecToVecTestParams<T> &p)
+{
+    const size_t sizes[] = { 1, 2, 3, 4, 0 };
+    RandomSeed seed(gRandomSeed);
+    float ulpConst = p.ulpLimit;
+
+    for (unsigned size = 0; sizes[size] != 0; size++)
+    {
+        p.ulpLimit = std::ceil(ulpConst + sizes[size]);
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_vec_to_vec.cpp]: Done. Correct, it is not needed.

**Review by EwanC — COMMENTED** [review]
I've looked over everything here now 🙂 

For the points highlighted in the PR description, I think the ULP errors and verification are correct. Nothing jumps out as inconsistent in how you've done the half tricky numbers, but I don't have
any knowledge about why we use the existing tricky numbers for float/double.

I've left a comment about TEST_SIZE though.

File: test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp [hunk 13]
```diff
@@ -0,0 +1,737 @@
...
+    float ulpConst = p.ulpLimit;
+
+    for (unsigned size = 0; sizes[size] != 0; size++)
+    {
+        p.ulpLimit = ulpConst + 2.f * sizes[size];
+
+        cl_int error = TwoVecToScalarKernel<T>(sizes[size], seed, p);
+        if (error != CL_SUCCESS)
+        {
+            log_error("   geom_two_args vector size %zu FAILED\n", sizes[size]);
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp]: Done

File: test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp [hunk 14]
```diff
@@ -0,0 +1,737 @@
...
+        value = (double)srcA[i] - (double)srcB[i];
+        total += value * value;
+    }
+
+    return sqrt(total);
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T> static double verifyLength(const T *srcA, size_t vecSize)
```
[the remark this answers is not in the document]
Comment (shajder) [reply test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp]: Done. Indeed I left duplicate as a trade off to keep such procedures local for consistency.

**Comment by bcalidas** [comment]
We need more time to review this PR.

File: test_conformance/geometrics/test_geometrics_fp.cpp (hunk 1 above)
Comment (EwanC) [reply test_conformance/geometrics/test_geometrics_fp.cpp]: Wrote this up as a OpenCL-Doc issue - https://github.com/KhronosGroup/OpenCL-Docs/issues/900

**Comment by lakshmih** [comment]
Looks good.

**Comment by lakshmih** [comment]
We would like to double check results at our end, can we get some more time.

**Comment by lakshmih** [comment]
We are opposed to this change merging in present form. The precision should be calculated in the same type as being used by the builtin (half for half, and float for float)

**Comment by EwanC** [comment]
> > We are opposed to this change merging in present form. The precision should be calculated in the same type as being used by the builtin (half for half, and float for float)
> 
> @lakshmih could you please point out specific fragment of code you are referring to ?

Qualcomm commented on the OpenCL-Docs issue I linked on my discussion above: https://github.com/KhronosGroup/OpenCL-CTS/pull/1633#discussion_r1155571948 So i'd assume the dot/cross verification code

File: test_conformance/geometrics/test_geometrics_cross.cpp [hunk 15]
```diff
@@ -0,0 +1,377 @@
...
+        : GeometricsFPTest(d, c, q)
+    {}
+    cl_int SetUp(int elements) override;
+    cl_int RunSingleTest(const GeomTestBase *p) override;
+
+    template <typename T> cl_int CrossKernel(const GeomTestParams<T> &p);
+};
+
+//--------------------------------------------------------------------------
+template <typename T>
```
Comment (lakshmih) [inline test_conformance/geometrics/test_geometrics_cross.cpp:57]: Would like the reference and error tolerance calculations in the two functions below to use the same type as the input, half for half and float for float as was the case earlier. 

Similarly for the reference and error tolerance calculation for dot product.

File: test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp [hunk 16]
```diff
@@ -0,0 +1,687 @@
...
+    if (std::is_same<T, half>::value)
+        return HTF(a) * HTF(b) + HTF(c);
+    else
+        return a * b + c;
+}
+
+//--------------------------------------------------------------------------
+
+template <typename T>
+double verifyDot(const T *srcA, const T *srcB, size_t vecSize)
```
Comment (lakshmih) [inline test_conformance/geometrics/test_geometrics_two_vec_to_scalar.cpp:153]: These reference calculations should use the same type as the input as well.

**Review by lakshmih — COMMENTED** [review]
Marked the code segments where the change is being requested.

**Comment by bashbaug** [comment]
See comments on the related issue:

* https://github.com/KhronosGroup/OpenCL-Docs/issues/900#issuecomment-1588601203
* https://github.com/KhronosGroup/OpenCL-Docs/issues/900#issuecomment-1602054509

In short, even our pure fp16 `dot` implementation isn't failing the new test, but a pure fp16 `cross` implementation is.  It would be great to get this clarified in the spec and to update the tests, also.

**Comment by lakshmih** [comment]
Could we remove focused review from this change since it is pending resolution of the spec discussion? We can add it back once we get there.

**Comment by bashbaug** [comment]
Removing "focused review" - we can revisit when the spec issues are resolved.
